### DISCUSSION TITLE
[done] Подчинённые таблицы в виде вкладок в документах.

### DISCUSSION THREAD

**Discussion by EKhorikov** [totumonline/totum-ru-issues-and-discussions#30] [Есть идея, а что если...]
Было бы неплохо, чтоб был базовый функционал с быстрыми настройками подчинённых таблиц.
Т.е. есть документ, при входе в который, открывается основной раздел и ряд вкладок с записями в подчинённые таблицы.
Например, это удобно, когда есть заказ на производство в ERP. Есть документ "Заказ на производство" в который входят изделия, которые необходимо произвести, а так же вкладки: "Материалы" в которой отображаются материалы используемые в производстве, "Задачи" в которой назначаются производственные задачи для сотрудников производства и т.д.

Сейчас в totum есть возможность создавать вкладки со связанными таблицами в циклах, либо создавать кнопки с переходом в связанные таблицы, но настройка и того и другого не является базовой.

На мой взгляд, это привычный интерфейс, по другим программным продуктам.

![image](https://user-images.githubusercontent.com/92679628/137638211-6dde92f8-3dba-4448-af2f-5cbeebe4de7f.png)

![image](https://user-images.githubusercontent.com/92679628/137638580-c5228a0e-e0b2-487b-a122-4d147575905a.png)

**Comment by ttm-alexeypolunin** [comment]
Добрый день. Эта задача была записана еще пару лет назад как «Ярлыки для циклов». Спасибо, что написали — нам стало понятно как это должно работать. Обсудил с разработкой — они говорят, что шанс на успех больше 60% (что классно).

### Как это будет выглядеть:

В таблицу циклов в строчную часть добавляются кнопки в которых ожидается возврат `linkToTable`, `linkToData***` и `linkToScript`.

Кнопки эти должны будут иметь специальный префикс в `name` — например `tab_users`. Каждая кнопка при открытии цикла будет преобразована во вкладку. Исполнятся она будет при клике на нее.

Далее мы открывает таблицу по выполненному `linkTo**` и встраиваем ее в обвязку цикла, те для пользователя это выглядит бесшовно. Точек привязки может быть несколько — все зависит от передаваемых префильтров.

Аналогично ярлыкам дерева, переданные префильтры будут заблокированы на изменение.

Мы думаем, что скорее всего можно будет даже перейти сразу на вкладку используя адрес типа `linkToTable(table: 'name_cycles_table.tab_name'; target: "self")`.

Также это решает задачу со специальными модулями, которые должны быть отображены внутри цикла — они будут вызываться через `linkToScript`. Тк модуль находится на том же хосте, то он сможет встраиваться в js тотума и быть бесшовно для пользователя отображен в теле цикла.

**Reply by ttm-alexeypolunin** [reply]
### Готово в .37 версии https://docs.totum.online/tables#type-tab-in-cyclea

`linkToScript` пока в таком режиме использовать нельзя!

**Reply by juramebel** [reply]
Здравствуйте.
Правильно я понимаю, эти вкладки не будут доступны пользователю, если у него доступ только на чтение этих таблиц?

**Reply by ttm-alexeypolunin** [reply]
Вкладки не будут доступны если он не видит кнопок с линками в таблице циклов — точнее он их там в принципе не видит, но если у них стоит `Видно ролям` и его роли там нет, то вкладка для него будет скрыта.

Еще может быть такая история, что у пользователя к таблице циклов с кнопками `tab_` доступ только на чтение. Получается что он нажать на кнопку не может и у него не буде вкладки. В этом случае кнопкам должны быть включены `Действие доступно при просмотре`.

**Reply by juramebel** [reply]
Точно, сработало.
Спасибо!

**Reply by mihail-cornev** [reply]
Подскажите как вы реализовали отображение других таблиц в т.циклов? Не пойму как это работает ((

**Reply by ttm-alexeypolunin** [reply]
https://ru.docs.totum.online/tables#type-tab-in-cyclea

Называется TAB

**Reply by mihail-cornev** [reply]
Я прочитал, но не понял как создать, к сожалению. Куда вставлять linktotable? Какой тип таблицы создавать в т.циклов и как туда привязать другую таблицу. 

Забыл сказать, что я пробовал добавить linktotable в код действия таблицы, но оно не работает - один раз открывает другую таблицу, а последующие уже не работает.

**Reply by juramebel** [reply]
Вам нужно в самой таблице циклов "снаружи" сделать кнопку, название которой обязательно начнется с префикса tab_
Например tab_1
В коде кнопки link на таблицу, которая нужна.
Например у меня
= :linkToTable(table: 'tehkarty'; filter: 'h_zakaz_na_proizvodstvo'=#id)

**Reply by mihail-cornev** [reply]
Как это снаружи? В хедере добавить кнопку и она потом встанет в ряд с табами?

**Reply by juramebel** [reply]
Нет. просто в таблице циклов. 
Я имел ввиду, что не в расчетной внутри, а в таблице циклов.

**Reply by mihail-cornev** [reply]
> Нет. просто в таблице циклов. Я имел ввиду, что не в расчетной внутри, а в таблице циклов.

  Спасибо, получилось. Наверное я единственный кто не понял по инструкции как это сделать. Я никак не мог взять в толк, что это должно быть именно в СТРОЧНОЙ части таблицы как отдельное поле.
